src/mumble/GlobalShortcut_linux.cpp

```diff
@@ -0,0 +1,187 @@
+#endif
+
+/**
+ * Returns a platform specific GlobalShortcutEngine object.
+ *
+ * @see GlobalShortcutX
+ * @see GlobalShortcutLinux
+ * @see GlobalShortcutMac
+ * @see GlobalShortcutWin
+ */
+GlobalShortcutEngine *GlobalShortcutEngine::platformInit() {
+	return new GlobalShortcutLinux();
+}
+
+GlobalShortcutLinux::GlobalShortcutLinux() {
+	if (g.s.bEnableEvdev) {
+		QString dir             = QLatin1String("/dev/input");
+		QFileSystemWatcher *fsw = new QFileSystemWatcher(QStringList(dir), this);
+		connect(fsw, SIGNAL(directoryChanged(const QString &)), this, SLOT(directoryChanged(const QString &)));
+		directoryChanged(dir);
+
+		if (qsKeyboards.isEmpty()) {
+			foreach (QFile *f, qmInputDevices)
+				delete f;
+			qmInputDevices.clear();
+
+			delete fsw;
+			qWarning("GlobalShortcutLinux: Unable to open any keyboard input devices under /dev/input");
+		} else {
+			return;
+		}
+	} else {
+		qWarning("GlobalShortcutLinux: No evdev support, global shortcuts disabled");
+	}
+}
+
+GlobalShortcutLinux::~GlobalShortcutLinux() {
+	wait();
+}
+
+// One of the raw /dev/input devices has ready input
+void GlobalShortcutLinux::inputReadyRead(int) {
+	if (!g.s.bEnableEvdev) {
+		return;
+	}
+
+	struct input_event ev;
+
+	if (bNeedRemap)
+		remap();
+
+	QFile *f = qobject_cast< QFile * >(sender()->parent());
+	if (!f)
+		return;
+
+	bool found = false;
+
+	while (f->read(reinterpret_cast< char * >(&ev), sizeof(ev)) == sizeof(ev)) {
+		found = true;
+		if (ev.type != EV_KEY)
+			continue;
+		bool down;
+		switch (ev.value) {
+			case 0:
+				down = false;
+				break;
+			case 1:
+				down = true;
+				break;
+			default:
+				continue;
+		}
+		int evtcode = ev.code + 8;
+		handleButton(evtcode, down);
+	}
+
+	if (!found) {
+		int fd      = f->handle();
+		int version = 0;
+		if ((ioctl(fd, EVIOCGVERSION, &version) < 0) || (((version >> 16) & 0xFF) < 1)) {
+			qWarning("GlobalShortcutLinux: Removing dead input device %s", qPrintable(f->fileName()));
+			qmInputDevices.remove(f->fileName());
+			qsKeyboards.remove(f->fileName());
+			delete f;
+		}
+	}
+}
+
+#define test_bit(bit, array) (array[bit / 8] & (1 << (bit % 8)))
+
+// The /dev/input directory changed
+void GlobalShortcutLinux::directoryChanged(const QString &dir) {
+	if (!g.s.bEnableEvdev) {
+		return;
+	}
+
+	QDir d(dir, QLatin1String("event*"), 0, QDir::System);
+	foreach (QFileInfo fi, d.entryInfoList()) {
+		QString path = fi.absoluteFilePath();
+		if (!qmInputDevices.contains(path)) {
+			QFile *f = new QFile(path, this);
+			if (f->open(QIODevice::ReadOnly)) {
+				int fd = f->handle();
+				int version;
+				char name[256];
+				uint8_t events[EV_MAX / 8 + 1];
+				memset(events, 0, sizeof(events));
+				if ((ioctl(fd, EVIOCGVERSION, &version) >= 0) && (ioctl(fd, EVIOCGNAME(sizeof(name)), name) >= 0)
+					&& (ioctl(fd, EVIOCGBIT(0, sizeof(events)), &events) >= 0) && test_bit(EV_KEY, events)
+					&& (((version >> 16) & 0xFF) > 0)) {
+					name[255] = 0;
```

Comment:
I would suggest to define the array's size as a constant above as 
```
constexpr unsigned int size = 256;
```
and then here you can do
```
name[size - 1] = 0;
```
That way we are safe when changing the array size in the future